lualib/resty/events/worker.lua

```diff
@@ -104,6 +125,98 @@ function _M.new(opts)
     return setmetatable(self, _MT)
 end
 
+local function read_thread(self, broker_connection)
+    while not terminating(self) do
+        local data, err = broker_connection:recv_frame()
+        if err then
+            if not is_timeout(err) then
+                return err_if_not_exiting("failed to read event", err)
+            end
+
+            -- timeout
+            goto continue
+        end
+
+        if not data then
+            log_if_not_exiting(ERR, "did not receive event from broker")
+            goto continue
+        end
+
+        local event_data, err = decode(data)
+        if err then
+            log_if_not_exiting(ERR, "failed to decode event data: ", err)
+            goto continue
+        end
+
+        -- got an event data, push to queue, callback in events_thread
+        local _, err = self._sub_queue:push(event_data)
+        if err then
+            log_if_not_exiting(ERR, "failed to store event: ", err, ". ",
+                    "data is: ", cjson_encode(event_data))
+        end
+
+        ::continue::
+    end -- while not terminating
+
+    return true
+end
+
+local function write_thread(self, broker_connection)
+    local counter = 0
+
+    while not terminating(self) do
+        local payload, err = self._pub_queue:pop()
+        if err then
+            if not is_timeout(err) then
+                return err_if_not_exiting("semaphore wait error", err)
+            end
+
+            -- timeout
+            goto continue
+        end
+
+        local _, err = broker_connection:send_frame(payload)
+        if err then
+            return err_if_not_exiting("failed to send event", err)
+        end
+
+        -- events rate limiting
+        counter = counter + 1
+        if counter >= EVENTS_COUNT_LIMIT then
+            sleep(EVENTS_SLEEP_TIME)
+            counter = 0
+        end
+
+        ::continue::
+    end -- while not terminating
+
+    return true
+end
+
+local function events_thread(self)
+    while not terminating(self) do
+        local data, err = self._sub_queue:pop()
+        if err then
+            if not is_timeout(err) then
+                return err_if_not_exiting("semaphore wait error", err)
+            end
+
+            -- timeout
+            goto continue
+        end
+
+        -- got an event data, callback
+        do_event(self, data)
+
+        -- yield, not block other threads
+        sleep(0)
```

Comment:
Seems we don't have an explicit yield in the `read_thread`?

Reply:
`read_thread` yields with: `broker_connection:recv_frame()`